pkg/hub/localizer/localizer.go

```diff
@@ -122,7 +133,282 @@ func (l *Localizer) handleGlobalization(glob *appsapi.Globalization) error {
+			if err != nil {
+				msg := fmt.Sprintf("failed to get Localizations matching for Description %s: %v", klog.KObj(desc), err)
+				klog.Error(msg)
+				l.recorder.Event(desc, corev1.EventTypeNormal, "FailedRetrievingLocalization", msg)
+				return err
+			}
+
+			if len(localizations) <= 0 {
+				klog.V(4).Info("cannot found Localizations matching for Description %s", klog.KObj(desc))
+			} else {
+				for _, loc := range localizations {
+					outputObject, err = l.executionOverride(appsapi.GenericOverride, loc.Spec.Overrides, outputObject)
+					if err != nil {
+						msg := fmt.Sprintf("failed to override for Description %s and Localization %s: %v", klog.KObj(desc), klog.KObj(loc), err)
+						klog.Error(msg)
+						l.recorder.Event(desc, corev1.EventTypeWarning, "FailedLocalizationObject", msg)
+						return err
+					}
+				}
+			}
+
+			outputObjects = append(outputObjects, outputObject)
+		}
+	}
+
+	desc.Spec.Raw = outputObjects
 	return nil
 }
+
+func (l *Localizer) getGlobalizationByResource(unstructured *unstructured.Unstructured) ([]*appsapi.Globalization, error) {
+	var glos []*appsapi.Globalization
+
+	globalizations, err := l.globLister.List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, glo := range globalizations {
+		if IsFeedMatchObject(glo.Spec.Feed, unstructured) {
+			glos = append(glos, glo)
+		}
+	}
+
+	return glos, nil
+}
+
+func (l *Localizer) getLocalizationByResource(namespace string, unstructured *unstructured.Unstructured) ([]*appsapi.Localization, error) {
+	var locs []*appsapi.Localization
+
+	localizations, err := l.locLister.Localizations(namespace).List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, loc := range localizations {
+		if IsFeedMatchObject(loc.Spec.Feed, unstructured) {
+			locs = append(locs, loc)
+		}
+	}
+
+	return locs, nil
+}
+
+func (l *Localizer) getGlobalizationByChartRef(chartRef appsapi.ChartReference) ([]*appsapi.Globalization, error) {
+	var glos []*appsapi.Globalization
+
+	globalizations, err := l.globLister.List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, glo := range globalizations {
+		if IsFeedMatchHelmChart(glo.Spec.Feed, chartRef) {
+			glos = append(glos, glo)
+		}
+	}
+
+	return glos, nil
+}
+
+func (l *Localizer) getLocalizationByChartRef(namespace string, chartRef appsapi.ChartReference) ([]*appsapi.Localization, error) {
+	var locs []*appsapi.Localization
+
+	localizations, err := l.locLister.Localizations(namespace).List(labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, loc := range localizations {
+		if IsFeedMatchHelmChart(loc.Spec.Feed, chartRef) {
+			locs = append(locs, loc)
+		}
+	}
+
+	return locs, nil
+}
+
+func (l *Localizer) executionOverride(overriderType appsapi.OverrideType, overrides []appsapi.OverrideRule, raw []byte) ([]byte, error) {
+	output := raw
+	var err error
+	for _, override := range overrides {
+		if overriderType == appsapi.GenericOverride {
```

Comment:
Could it be better we could aggregate all the overrides, and then apply final override once.

Reply:
Customer may use different patch strategy in the same Localization. for each override item, we need to parser the values then know the patch part is a JsonPatch or a SMP Patch, so execute override one by one should be easier here.
`apiVersion: apps.clusternet.io/v1alpha1
kind: Localization
metadata:
  name: foo-loc
  namespace: clusternet-px5n9
spec:
  overrides:
    - ruleName: addLabel2
      overrideValue:
        patch: |-
          apiVersion: v1
          kind: Namespace
          metadata:
            name: foo
            labels:
              clusternet.io/localization-injected: true
    - ruleName: addAnno
      overrideValue:
        patch: |-
          - path: "/metadata/annotations"
            op: "add"
            value:
              app: apple
  feed:
    apiVersion: v1
    kind: Namespace
    name: foo`

Reply:
It's more clear to aggregate HelmChart values first and merge values finally.